s3/s3_operations.cpp

```diff
@@ -0,0 +1,1682 @@
+                                std::string bucket_name = tokens.front();
+                                std::string object_name = tokens.back();
+
+                                // reverse object id
+                                std::string obj_id = boost::lexical_cast<std::string>(object->id());
+                                std::reverse(obj_id.begin(), obj_id.end());
+
+                                // make S3 key name
+                                std::ostringstream s3_key_name;
+                                s3_key_name << "/" << bucket_name << "/" << obj_id << "/" << object_name;
+
+                                // update physical path
+                                object->physical_path(s3_key_name.str());
+                            }
+
+                            ret = s3PutCopyFile(S3_PUTFILE, _cache_file_name, object->physical_path(), statbuf.st_size, key_id, access_key, _ctx.prop_map());
+                            result = ASSERT_PASS(ret, "[resource_name=%s] Failed to copy the cache file: \"%s\" to the S3 object: \"%s\".", get_resource_name(_ctx.prop_map()).c_str(),
+                                                 _cache_file_name, object->physical_path().c_str());
+
+                        }
+                    }
+                }
+            }
+            if( !result.ok() ) {
+                std::stringstream msg;
+                msg << "[resource_name=" << get_resource_name(_ctx.prop_map()) << "] "
+                    << result.result();
+                rodsLog(LOG_ERROR, msg.str().c_str());
+            }
+            return result;
+        }
+
+    }
+
+    // =-=-=-=-=-=-=-
+    // used to allow the resource to determine which host
+    // should provide the requested operation
+    irods::error s3_resolve_resc_hier_operation(
+        irods::plugin_context& _ctx,
+        const std::string*                  _opr,
+        const std::string*                  _curr_host,
+        irods::hierarchy_parser*            _out_parser,
+        float*                              _out_vote )
+    {
+        rodsLog(debug_log_level, "%s:%d (%s) [[%lu]] _opr=%s\n", __FILE__, __LINE__, __FUNCTION__, std::hash<std::thread::id>{}(std::this_thread::get_id()), _opr->c_str());
+
+        irods::file_object_ptr file_obj = boost::dynamic_pointer_cast<irods::file_object>(_ctx.fco());
+
+        // read the data size and save and save it under the property plugin property map
+        uint64_t data_size = 0;
+        char *data_size_str = getValByKey(&file_obj->cond_input(), DATA_SIZE_KW);
+        if (data_size_str) {
+            try {
+                data_size = boost::lexical_cast<uint64_t>(data_size_str);
+            } catch (boost::bad_lexical_cast const& e) {
+                data_size = 0;
+            }
+        }
+        {
+            std::lock_guard<std::mutex> lock(global_mutex);
+            irods_s3::data_size = data_size;
+        }
```

Comment:
are these extra brackets necessary?   if so, please explain in a comment.

Reply:
Yes, to control the scope of the lock_guard.  Comment added.